packages/eslint-plugin/src/rules/no-unsafe-declaration-merging.ts

```diff
@@ -0,0 +1,52 @@
+import type { TSESTree } from '@typescript-eslint/utils';
+import * as ts from 'typescript';
+
+import * as util from '../util';
+
+export default util.createRule({
+  name: 'no-unsafe-declaration-merging',
+  meta: {
+    type: 'problem',
+    docs: {
+      description: 'Disallow unsafe declaration merging',
+      recommended: 'strict',
+      requiresTypeChecking: true,
```

Comment:
We should actually be able to do this without type info and just with the scope analysis APIs.

![image](https://user-images.githubusercontent.com/7462525/196681607-6c635160-a47b-4746-842c-636d303be9b4.png)
[playground](https://typescript-eslint.io/play/#ts=4.8.4&sourceType=module&showAST=scope&code=JYOwLgpgTgZghgYwgAgGIHt3IN7IL4CwAUAgDZwDOFamO+QA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA)

You can see that if the names clash, then we can iterate through the `variable.defs` array and do a similar comparison as you've done here - `variable.defs.some(def => def.node.type === unsafeType)` where
`unsafeType == AST_NODE_TYPES.TSInterfaceDeclaration | AST_NODE_TYPES.ClassDeclaration`

Reply:
Ah, I should finally get familiar with the scope manager 😄. #5854

Reply:
@yeonjuan do you have time to tackle this soon? I can if not, no worries 😄

Reply:
Hi @bradzacher, @JoshuaKGoldberg  I can fix this soon. I'll make a PR for it. (including [fixing typo](https://github.com/typescript-eslint/typescript-eslint/pull/5840#discussion_r999326647))
 Thanks!